Support GRBM training with hidden units

kevinchern opened the pull request:
Addressing https://github.com/dwavesystems/dwave-pytorch-plugin/issues/7

mhramani reviewed (changes requested):
Thanks for the PR, I made a quick initial review. Will make another one later.

thisac commented on tests/test_boltzmann_machine.py:
```diff
@@ -78,6 +88,43 @@ def test_forward(self):
             en_boltz = self.bm(fake_spins).item()
             self.assertAlmostEqual(en_bqm, en_boltz, 4)
 
+    def test_estimate_beta(self):
```
This could use some parameterization (worth including [parameterized](https://pypi.org/project/parameterized/) as a dev dependency IMO).

thisac commented on tests/test_boltzmann_machine.py:
```diff
@@ -78,6 +88,43 @@ def test_forward(self):
+        corrupted = torch.tensor([[-1, 1, torch.nan]])
+        expected = grbm.compute_expectation_disconnected(corrupted, beta)[0].tolist()
+        self.assertListEqual(expected[:2], [-1, 1])
+        # -0.13 - 0.17 + 0.3 = 0.1
+        # 0.1 * 1.337 = 0.1337
+        # -tanh(0.1337)
+        self.assertAlmostEqual(expected[-1], -0.132909)
```
Pardon my ignorance, but where does `-0.13 - 0.17 + 0.3 = 0.1` come from? Any reason not to do the calculations in code instead of hardcoding it?

kevinchern replied:
The expression is derived from the example constructed a few lines above. I'm not sure what the best-practice is so I hard-coded it. I updated the documentation to make the test case more clear.

thisac reviewed (commented):
Thanks @kevinchern.

kevinchern commented:
Thanks for the feedback @mhramani @thisac! I vectorized the computation, simplified the usage (and example), improved documentation, and added more tests.

kevinchern commented:
The two examples `examples/boltzmann_machine_hidden.py` and `examples/boltzmann_machine.py` share a lot of duplicate code. Should we merge the two examples instead?

mhramani commented on dwave/plugins/torch/boltzmann_machine.py:
```diff
@@ -87,6 +97,79 @@ def sufficient_statistics(
+        """
+        m = self._pad(obs)
+        h_eff = self._compute_effective_field(m)
+        m[:, self.hidx] = -torch.tanh(h_eff * beta)
```
shouldn't we sample hidden units using torch.bernoulli once we calculate their probabilities?

mhramani commented on dwave/plugins/torch/boltzmann_machine.py:
```diff
@@ -214,6 +344,38 @@ def interactions(self, x: torch.Tensor) -> torch.Tensor:
         """
         return x[..., self.edge_idx_i] * x[..., self.edge_idx_j]
 
+    def _compute_effective_field(self, padded: torch.Tensor) -> torch.Tensor:
```
Should we define the device that runs this code?

mhramani reviewed (commented):
Looks good, left 2 comment / questions in the code.
